File: server/parser.go
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"github.com/fatih/color"
	"github.com/rodaine/table"
	"strconv"
)

var CommandMap = map[string]uint32{
	"dir":      CommandDir,
	"mods":     CommandMods,
	"shutdown": CommandShutdown,
}

var TableMap = map[uint32]TableHeaders{
	TypeCheckin: {
		Headers: []string{"PeerId", "Hostname", "Domain", "Username", "Interfaces"},
		Values:  make([]string, 5),
	},
	CommandDir: {
		Headers: []string{"Mode", "Length", "LastWriteTime", "Name"},
		Values:  make([]string, 4),
	},
	CommandMods: {
		Headers: []string{"ModName", "BaseAddress"},
		Values:  make([]string, 2),
	},
}

func CreateParser(buffer []byte) *Parser {
	var parser = new(Parser)

	parser.Buffer = buffer
	parser.BigEndian = true
	parser.Length = uint32(len(buffer))

	return parser
}

func (p *Parser) ParseByte() []byte {
	var buffer = make([]byte, 1)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.Length >= 1 {

		copy(buffer, p.Buffer[:1])
		p.Length -= 1

		if p.Length == 0 {
			p.Buffer = []byte{}

		} else {
			p.Buffer = p.Buffer[1:]
		}
	}

	return buffer
}

func (p *Parser) ParseBool() bool {
	var integer = make([]byte, 4)

	for i := range integer {
		integer[i] = 0
	}

	if p.Length >= 4 {

		copy(integer, p.Buffer[:4])
		p.Length -= 4

		if p.Length == 0 {
			p.Buffer = []byte{}

		} else {
			p.Buffer = p.Buffer[4:]
		}
	}

	if p.BigEndian {
		return int(binary.BigEndian.Uint32(integer)) != 0

	} else {
		return int(binary.LittleEndian.Uint32(integer)) != 0
	}
}

func (p *Parser) ParseDword() uint32 {
	var buffer = make([]byte, 4)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.Length >= 4 {

		copy(buffer, p.Buffer[:4])
		p.Length -= 4

		if p.Length == 0 {
			p.Buffer = []byte{}

		} else {
			p.Buffer = p.Buffer[4:]
		}
	}

	if p.BigEndian {
		return binary.BigEndian.Uint32(buffer)

	} else {
		return binary.LittleEndian.Uint32(buffer)
	}
}

func (p *Parser) ParseDword64() uint64 {
	var buffer = make([]byte, 8)

	for i := range buffer {
		buffer[i] = 0
	}

	if p.Length >= 8 {

		copy(buffer, p.Buffer[:8])
		p.Length -= 8

		if p.Length == 0 {
			p.Buffer = []byte{}
		} else {
			p.Buffer = p.Buffer[8:]
		}
	}

	if p.BigEndian {
		return binary.BigEndian.Uint64(buffer)

	} else {
		return binary.LittleEndian.Uint64(buffer)
	}
}

func (p *Parser) ParseBytes() []byte {
	var buffer []byte

	if p.Length >= 4 {

		size := p.ParseDword()
		fmt.Printf("length of bytes: %d\n", size)

		if size != 0 {
			buffer = make([]byte, size)
			copy(buffer, p.Buffer[:size])

			p.Length -= size

			if p.Length == 0 {
				p.Buffer = []byte{}

			} else {
				p.Buffer = p.Buffer[size:]
			}
		}
	}

	return buffer
}

func (p *Parser) ParseWString() string {
	return string(bytes.Trim([]byte(DecodeUTF16(p.ParseBytes())), "\x00"))
}

func (p *Parser) ParseString() string {
	return string(p.ParseBytes())
}

func (p *Parser) ParserPrintData(TypeId uint32) bool {
	var (
		tMap TableHeaders
		ok   bool
	)

	if TypeId == TypeCheckin {
		if tMap, ok = TableMap[TypeCheckin]; ok {

			tMap.Values[0] = strconv.Itoa(int(p.PeerId))
			tMap.Values[1] = p.ParseString()
			tMap.Values[2] = p.ParseString()
			tMap.Values[3] = p.ParseString()
			tMap.Values[4] = p.ParseString()
		}
	}

	if TypeId == TypeTasking {
		CmdId := p.ParseDword()

		if CmdId == CommandDir {
			fmt.Println("================ CommandDir ================")

			if tMap, ok = TableMap[CommandDir]; ok {

				for p.Length != 0 {
					IsDir := p.ParseBool()

					if IsDir == true {
						fmt.Println("entry is a directory...")
						tMap.Values[0] = "dir"
						tMap.Values[1] = "n/a"

					} else {
						fmt.Println("entry is a file. parsing size...")
						tMap.Values[0] = ""
						tMap.Values[1] = strconv.Itoa(int(p.ParseDword64()))
						fmt.Printf("file size: %s\n", tMap.Values[1])
					}

					Day := p.ParseDword()
					Month := p.ParseDword()
					Year := p.ParseDword()

					Second := p.ParseDword()
					Minute := p.ParseDword()
					Hour := p.ParseDword()

					tMap.Values[2] = fmt.Sprintf("%d/%d/%d %d:%d:%d", Day, Month, Year, Hour, Minute, Second)

					fmt.Println(tMap.Values[2])

					tMap.Values[3] = p.ParseString()
					fmt.Printf("name: %s\n", tMap.Values[3])
				}
			}
		}

		if CmdId == CommandMods {
			if tMap, ok = TableMap[CommandMods]; ok {

				tMap.Values[0] = p.ParseString()
				tMap.Values[1] = fmt.Sprintf("0x%X", p.ParseDword64())
			}
		}
	}

	return ParseTable(tMap)
}

func ParseTable(tmap TableHeaders) bool {

	hdrs := tmap.Headers
	vals := tmap.Values

	for k, v := range vals {
		if v == "" {
			vals[k] = "null"
		}
	}

	headersInterface := make([]interface{}, len(hdrs))
	for i, header := range hdrs {
		fmt.Printf("creating interface for header \"%s\"\n", header)
		headersInterface[i] = header
	}

	valuesInterface := make([]interface{}, len(vals))
	for i, val := range vals {
		fmt.Printf("creating interface for value \"%s\"\n", val)
		valuesInterface[i] = val
	}

	fmt.Println("creating new table")
	tbl := table.New(headersInterface...)
	format := color.New(color.FgCyan).SprintfFunc()

	fmt.Println("setting table formatter")
	tbl.WithHeaderFormatter(format)
	fmt.Printf("adding row %s\n", valuesInterface)
	tbl.AddRow(valuesInterface...)

	fmt.Println()
	fmt.Println("printing...")
	tbl.Print()
	fmt.Println()

	return true
}

File: server/src/parser.rs
use byteorder::{BigEndian, LittleEndian, ByteOrder};
use std::convert::TryInto;

pub struct Parser {
    msg_buffer: Vec<u8>,
    pointer:    usize,
    big_endian: bool,
    msg_length: u32,
    peer_id:    u32,
    task_id:    u32,
    msg_type:   u32,
}
impl Parser {
    pub fn create_parser(buffer: Vec<u8>) -> Parser {
        let mut parser = Parser {
            msg_buffer: buffer,
            big_endian: true,
            msg_length: 0,
            pointer:    0,
            peer_id:    0,
            task_id:    0,
            msg_type:   0,
        };

        parser.msg_length   = parser.msg_buffer.len() as u32;
        parser.peer_id      = parser.parse_uint32();
        parser.task_id      = parser.parse_uint32();
        parser.msg_type     = parser.parse_uint32();

        parser
    }

    pub fn parse_byte(&mut self) -> Vec<u8> {
        if self.msg_length >= 1 {
            let buffer = self.msg_buffer[self.pointer..self.pointer+1].to_vec();

            self.pointer += 1;
            self.msg_length -= 1;

            buffer
        }
        else {
            vec![0; 1]
        }
    }

    pub fn parse_bool(&mut self) -> bool {
        let integer = self.parse_uint32();
        integer != 0
    }

    pub fn parse_uint32(&mut self) -> u32 {
        if self.msg_length >= 4 {
            let buffer = &self.msg_buffer[self.pointer..self.pointer+4].to_vec();

            self.pointer += 4;
            self.msg_length -= 4;

            if self.big_endian {
                BigEndian::read_u32(buffer)
            }
            else {
                LittleEndian::read_u32(buffer)
            }
        }
        else { 0 }
    }

    pub fn parse_uint3264(&mut self) -> u64 {
        if self.msg_length >= 8 {
            let buffer = &self.msg_buffer[self.pointer..self.pointer+8].to_vec();

            self.pointer += 8;
            self.msg_length -= 8;

            if self.big_endian {
                BigEndian::read_u64(buffer)
            }
            else {
                LittleEndian::read_u64(buffer)
            }
        }
        else { 0 }
    }

    pub fn parse_bytes(&mut self) -> Vec<u8> {
        let size = self.parse_uint32() as usize;

        if size > 0 && self.msg_length as usize >= size {
            let buffer = self.msg_buffer[self.pointer..self.pointer+size].to_vec();

            self.pointer += size;
            self.msg_length -= size as u32;

            buffer
        }
        else { vec![] }
    }

    pub fn parse_wstring(&mut self) -> String {
        let bytes = self.parse_bytes();
        let string = decode_utf16(&bytes);

        string.trim_end_matches('\x00').to_string()
    }

    pub fn parse_string(&mut self) -> String {
        String::from_utf8(self.parse_bytes()).unwrap_or_default()
    }
}

fn decode_utf16(bytes: &[u8]) -> String {
    let mut u16_vec = Vec::with_capacity(bytes.len() / 2);

    for chunk in bytes.chunks(2) {
        let arr: [u8; 2] = chunk.try_into().unwrap_or([0, 0]);
        u16_vec.push(u16::from_le_bytes(arr));
    }
    String::from_utf16_lossy(&u16_vec)
}
